File: 3186. Maximum Total Damage With Spell Casting.go
func maximumTotalDamage(power []int) int64 {
    sort.Ints(power)
    freq := make(map[int]int)
    for _, i := range power {
        freq[i] += 1
    }
    cache := make([]int64, len(power))
    for i := range cache {
        cache[i] = -1
    }
    return find(power, 0, freq, cache)
}
func find(power []int, index int, freq map[int]int, cache []int64) int64 {
    if index >= len(power) { return 0 }
    if cache[index] != -1 { return cache[index] }
    answer1 := find(power, index + freq[power[index]], freq, cache)
    answer2 := int64(power[index] * freq[power[index]])
    i := index + freq[power[index]] + freq[power[index]+1] + freq[power[index]+2]
    answer2 += find(power, i, freq, cache)
    cache[index] = max(answer1, answer2)
    return cache[index]
}

File: java/3186. Maximum Total Damage With Spell Casting.java
class Solution {
    public long maximumTotalDamage(int[] power) {
        Arrays.sort(power);
        Map<Integer, Integer> freq = new HashMap<>();
        for (int p : power) freq.put(p, freq.getOrDefault(p, 0)+1);
        Long[] dp = new Long[power.length];
        return find(power, 0, dp, freq);
    }
    private long find(int[] power, int index, Long[] dp, Map<Integer, Integer> freq) {
        if (index >= power.length) return 0;
        if (dp[index] != null) return dp[index];
        // don't take this element
        long answer1 = find(power, nextIndex(power, power[index]+1), dp, freq);
        // take this element
        long answer2 = (long)freq.get(power[index]) * power[index];
        answer2 += find(power, nextIndex(power, power[index]+3), dp, freq);
        return dp[index] = Math.max(answer1, answer2);
    }
    private int nextIndex(int[] power, int target) {
        int left = 0, right = power.length-1;
        int answer = power.length;
        while (left <= right) {
            int mid = left + (right-left)/2;
            if (power[mid] >= target) {
                answer = mid;
                right = mid-1;
            } else {
                left = mid+1;
            }
        }
        return answer;
    }
}
